File: system-test/java/src/main/java/com/optivem/shop/testkit/core/usecase/shop/usecases/CancelOrder.java
package com.optivem.shop.testkit.core.usecase.shop.usecases;

import com.optivem.shop.testkit.driver.port.shop.ShopDriver;
import com.optivem.shop.testkit.core.usecase.shop.usecases.base.BaseShopUseCase;
import com.optivem.shop.testkit.core.shared.UseCaseResult;
import com.optivem.shop.testkit.core.shared.UseCaseContext;
import com.optivem.shop.testkit.core.shared.VoidVerification;

public class CancelOrder extends BaseShopUseCase<Void, VoidVerification> {
    private String orderNumberResultAlias;

    public CancelOrder(ShopDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public CancelOrder orderNumber(String orderNumberResultAlias) {
        this.orderNumberResultAlias = orderNumberResultAlias;
        return this;
    }

    @Override
    public UseCaseResult<Void, VoidVerification> execute() {
        var orderNumber = context.getResultValue(orderNumberResultAlias);
        var result = driver.cancelOrder(orderNumber);
        return new UseCaseResult<>(result, context, VoidVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/shop/usecases/CancelOrder.ts
import type { ShopDriver } from '../../../../../driver/port/shop/shop-driver.js';
import { UseCaseResult } from '../../../shared/use-case-result.js';
import { VoidVerification } from '../../../shared/void-verification.js';
import type { UseCaseContext } from '../../../shared/use-case-context.js';
import { BaseShopUseCase } from './base/BaseShopUseCase.js';

export class CancelOrder extends BaseShopUseCase<void, VoidVerification> {
  private _orderNumberResultAlias: string = '';

  constructor(driver: ShopDriver, context: UseCaseContext) {
    super(driver, context);
  }

  orderNumber(orderNumberResultAlias: string): this {
    this._orderNumberResultAlias = orderNumberResultAlias;
    return this;
  }

  async execute(): Promise<UseCaseResult<void, VoidVerification>> {
    const orderNumber = this.context.getResultValue(this._orderNumberResultAlias) ?? this._orderNumberResultAlias;
    const result = await this.driver.cancelOrder(orderNumber);

    return new UseCaseResult(
      result,
      this.context,
      (_, ctx) => new VoidVerification(undefined, ctx),
    );
  }
}
